icu4c/source/i18n/measfmt.cpp

```diff
@@ -770,71 +715,96 @@ static void appendRange(
+    } else { // h m s, handled outside formatNumeric. No value is also an error.
         status = U_INTERNAL_PROGRAM_ERROR;
         return appendTo;
     }
 
-    // Format time. draft becomes something like '5:30:45'
-    // #13606: DateFormat is not thread-safe, but MeasureFormat advertises itself as thread-safe.
-    FieldPosition smallestFieldPosition(smallestField);
-    UnicodeString draft;
-    static UMutex dateFmtMutex;
-    umtx_lock(&dateFmtMutex);
-    dateFmt.format(date, draft, smallestFieldPosition, status);
-    umtx_unlock(&dateFmtMutex);
-
-    // If we find field for smallest amount replace it with the formatted
-    // smallest amount from above taking care to replace the integer part
-    // with what is in original time. For example, If smallest amount
-    // is 9.35s and the formatted time is 0:00:09 then 9.35 becomes 09.35
-    // and replacing yields 0:00:09.35
-    if (smallestFieldPosition.getBeginIndex() != 0 ||
-            smallestFieldPosition.getEndIndex() != 0) {
-        appendRange(draft, 0, smallestFieldPosition.getBeginIndex(), appendTo);
-        appendRange(
-                smallestAmountFormatted,
-                0,
-                intFieldPosition.getBeginIndex(),
-                appendTo);
-        appendRange(
-                draft,
-                smallestFieldPosition.getBeginIndex(),
-                smallestFieldPosition.getEndIndex(),
-                appendTo);
-        appendRange(
-                smallestAmountFormatted,
-                intFieldPosition.getEndIndex(),
-                appendTo);
-        appendRange(
-                draft,
-                smallestFieldPosition.getEndIndex(),
-                appendTo);
-    } else {
-        appendTo.append(draft);
+    const NumberFormat *numberFormatter = numberFormat->get();
+    auto* numberFormatter2 = (NumberFormat*) numberFormatter->clone();
+    numberFormatter2->setMinimumIntegerDigits(2);
+
+    FormattedStringBuilder fsb;
+
+    UBool protect = FALSE;
+    const int32_t patternLength = pattern.length();
+    for (int32_t i = 0; i < patternLength; i++) {
+        char c = pattern[i];
+
+        // Also set the proper field in this switch
+        // We don't use DateFormat.Field because this is not a date / time, is a duration.
+        double value = 0;
+        switch (c) {
+            case 'H': value = hours; break;
+            case 'm': value = minutes; break;
+            case 's': value = seconds; break;
+        }
+
+        // There is not enough info to add Field(s) for the unit because all we have are plain
+        // text patterns. For example in "21:51" there is no text for something like "hour",
+        // while in something like "21h51" there is ("h"). But we can't really tell...
+        switch (c) {
+            case 'H':
+            case 'm':
+            case 's':
+                if (protect) {
+                    fsb.append(UnicodeString(c), (FormattedStringBuilder::Field) NULL, status);
```

Comment:
For undefined field, please use

`UNUM_FIELD_COUNT`

I know, yuck yuck.  Why does UNUM_FIELD_COUNT mean null field when this is not a number-specific format?  For historical reasons.  I filed a bug to clean this up:

https://unicode-org.atlassian.net/browse/ICU-20665

Also, please use `appendCodePoint` when you are appending a single character.

Reply:
Done UNUM_FIELD_COUNT.

Can't use appendCodePoint, since the character is char16_t, so it might be a utf-16 surrogate, not a code point.

Reply:
Done